server/src-lib/Hasura/Server/Auth/WebHook.hs

```diff
@@ -0,0 +1,150 @@
+import           Hasura.HTTP
+import           Hasura.Logging
+import           Hasura.Prelude
+import           Hasura.RQL.Types
+import           Hasura.Server.Logging
+import           Hasura.Server.Utils
+
+
+data AuthHookType
+  = AHTGet
+  | AHTPost
+  deriving (Eq)
+
+instance Show AuthHookType where
+  show AHTGet  = "GET"
+  show AHTPost = "POST"
+
+isPost :: AuthHookType -> Bool
+isPost AHTGet  = False
+isPost AHTPost = True
+
+
+data AuthHookG a b
+  = AuthHookG
+  { ahUrl  :: !a
+  , ahType :: !b
+  } deriving (Show, Eq)
+
+type AuthHook = AuthHookG T.Text AuthHookType
+
+withAuthHook :: AuthHook -> (T.Text -> a) -> (T.Text -> a) -> a
+withAuthHook (AuthHookG u t) ifGet ifPost =
+  bool (ifGet u) (ifPost u) $ isPost t
+
+hookMethod :: AuthHook -> N.StdMethod
+hookMethod = bool N.GET N.POST . isPost . ahType
+
+
+-- | Extracts 'UserInfo' from the webhook answer.
+--   This uses the provided manager to make the call, extracts
+--   information such as response body and status, and delegates to a
+--   local 'mkUserInfoFromResp'.
+userInfoFromAuthHook
+  :: (HasVersion, MonadIO m, MonadError QErr m)
+  => Logger Hasura
+  -> H.Manager
+  -> AuthHook
+  -> [N.Header]
+  -> m (UserInfo, Maybe UTCTime)
+userInfoFromAuthHook logger manager hook reqHeaders = do
+  res  <- liftIO $ try $ withAuthHook hook withGET withPOST
+  resp <- either logAndThrow return res
+  let status   = resp ^. Wreq.responseStatus
+      respBody = resp ^. Wreq.responseBody
+  mkUserInfoFromResp logger (ahUrl hook) (hookMethod hook) status respBody
+
+  where
+    withGET  url = Wreq.getWith  (mkOptions filteredHeaders) $ T.unpack url
+    withPOST url = Wreq.postWith (mkOptions [contentType]) (T.unpack url) $
+      object ["headers" J..= postHdrsPayload]
+    mkOptions = wreqOptions manager
+    contentType = ("Content-Type", "application/json")
+    postHdrsPayload = J.toJSON $ Map.fromList $ hdrsToText reqHeaders
+
+    logAndThrow err = do
+      unLogger logger $
+        WebHookLog LevelError Nothing (ahUrl hook) (hookMethod hook)
+        (Just $ HttpException err) Nothing Nothing
+      throw500 "Internal Server Error"
+
+    filteredHeaders = flip filter reqHeaders $ \(n, _) ->
+      n `notElem` commonClientHeadersIgnored
+
+
+-- | Processes the result of the wehook call, and tries to construct a
+--   'UserInfo'. The expiration time is retrieved from either the
+--   CacheControl or Expires header.
+mkUserInfoFromResp
+  :: (MonadIO m, MonadError QErr m)
+  => Logger Hasura
+  -> T.Text
+  -> N.StdMethod
+  -> N.Status
+  -> BL.ByteString
+  -> m (UserInfo, Maybe UTCTime)
+mkUserInfoFromResp (Logger logger) url method statusCode respBody
+  | statusCode == N.status200 =
+    case eitherDecode respBody of
+      Left e -> do
+        logError
+        throw500 $ "Invalid response from authorization hook: " <> T.pack e
+      Right rawHeaders -> getUserInfoFromHdrs rawHeaders
+
+  | statusCode == N.status401 = do
+    logError
```

Comment:
Do we need to log an error here? If the authentication hook responds with 401, that seems like it’s just operating as intended, not an error (the error is on the user’s part, not the server). (This is a genuine question, I’m not 100% sure I understand what this functionality does.)

Reply:
That's a good point... but there again, this is existing code, that I didn't want to touch. Thinking about it: we are going to receive a 401 every time authentication fails... perhaps a warning is enough?